### PR TITLE
Local autogptq fix

### PR DESCRIPTION
**Pull request by achew010**
## Description

This PR applies additional fixes to #48,

### Fixed items:
- removed direct dependency in plugin
- edited plugin README with information to install direct dependencies when creating a legacy GPTQ-LORA
- Shifted `use_external_lib` argument from plugin argument to configuration object field
- Added additional unit tests to ensure 
    - the new field is read
    - the correct package is imported
    - appropriate errors are thrown
- regenerated sample configuration yamls to include the new field

### FMS-Acceleration-Peft Unit Test
```
============================================================================================================================================================================================================ test session starts ============================================================================================================================================================================================================
platform linux -- Python 3.10.12, pytest-8.2.2, pluggy-1.5.0
rootdir: /data/aaron/experimental/fms-acceleration/plugins/accelerated-peft
configfile: pyproject.toml
collected 8 items                                                                                                                                                                                                                                                                                                                                                                                                                           

tests/test_gptqmodel.py ..                                                                                                                                                                                                                                                                                                                                                                                                            [ 25%]
tests/test_peft_plugins.py ...                                                                                                                                                                                                                                                                                                                                                                                                        [ 62%]
tests/test_q4_triton.py ..                                                                                                                                                                                                                                                                                                                                                                                                            [ 87%]
tests/test_triton.py .                                                                                                                                                                                                                                                                                                                                                                                                                [100%]

============================================================================================================================================================================================================= warnings summary ==============================================================================================================================================================================================================
.tox/py/lib/python3.10/site-packages/transformers/utils/hub.py:124
  /data/aaron/experimental/fms-acceleration/plugins/accelerated-peft/.tox/py/lib/python3.10/site-packages/transformers/utils/hub.py:124: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.
    warnings.warn(

tests/test_gptqmodel.py::test_pre_quantized_model_outputs_match
tests/test_gptqmodel.py::test_quantizing_pretrained_model_outputs_match
tests/test_gptqmodel.py::test_quantizing_pretrained_model_outputs_match
tests/test_peft_plugins.py::test_autogptq_loading
tests/test_q4_triton.py::TestsQ4Triton::test_generation_desc_act_false
tests/test_q4_triton.py::TestsQ4Triton::test_generation_desc_act_true
tests/test_triton.py::TestTriton::test_triton_qlinear
  /data/aaron/experimental/fms-acceleration/plugins/accelerated-peft/.tox/py/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
    warnings.warn(

tests/test_gptqmodel.py::test_pre_quantized_model_outputs_match
  /data/aaron/experimental/fms-acceleration/plugins/accelerated-peft/.tox/py/lib/python3.10/site-packages/auto_gptq/utils/peft_utils.py:360: UserWarning: You can just ignore this warning if the peft type you use isn't in ['LORA', 'ADALORA'].
  LlamaGPTQForCausalLM supports injecting fused attention but not enables this time. If you are training adapters, you must also disable fused attention injection when loading quantized base model at inference time, otherwise adapters may not be added to base model properly. If you are loading adapters to do inference, you can reference to adapter's config file to check whether the adapters are trained using base model that n
ot enable fused attention injection.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================================================================================================================= 8 passed, 9 warnings in 140.29s (0:02:20) =================================================================================================================================================================================================
```

### FMS-HF-Tuning Acceleration Unit Test
```
============================================================================================================================================================================================================= warnings summary ==============================================================================================================================================================================================================
../fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/transformers/utils/hub.py:124
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/transformers/utils/hub.py:124: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_raises_due_to_invalid_arguments[triton_v2 requires fp16]
tests/acceleration/test_acceleration_framework.py::test_framework_raises_due_to_invalid_arguments[accelerated peft requires peft config]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/transformers/training_args.py:1847: FutureWarning: `--push_to_hub_token` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--hub_token` instead.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/huggingface_hub/utils/_deprecation.py:100: FutureWarning: Deprecated argument(s) used in '__init__': dataset_text_field, max_seq_length. Will not be supported from version '1.0.0'.
  
  Deprecated positional argument(s) used in SFTTrainer, please use the SFTConfig to set these arguments instead.
    warnings.warn(message, FutureWarning)

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/trl/trainer/sft_trainer.py:269: UserWarning: You passed a `max_seq_length` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/trl/trainer/sft_trainer.py:307: UserWarning: You passed a `dataset_text_field` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/trl/trainer/sft_trainer.py:397: UserWarning: You passed a tokenizer with `padding_side` not equal to `right` to the SFTTrainer. This might lead to some unexpected behaviour due to overflow issues when training a model in half-precision. You might consider adding `tokenizer.padding_side = 'right'` to your code.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[bitsandbytes]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_peft[auto_gptq]
tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-acceleration/.tox/run-benches/lib/python3.10/site-packages/accelerate/accelerator.py:447: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
  dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-hf-tuning/tuning/config/acceleration_configs/acceleration_framework_config.py:244: UserWarning: An experimental acceleration feature is requested by specifying the '--fused_lora' argument. Please note this feature may not support certain edge cases at this juncture. When the feature matures this message will be turned off.
    warnings.warn(

tests/acceleration/test_acceleration_framework.py::test_framework_intialized_properly_foak
  /data/aaron/experimental/fms-hf-tuning/tuning/config/acceleration_configs/acceleration_framework_config.py:244: UserWarning: An experimental acceleration feature is requested by specifying the '--fast_kernels' argument. Please note this feature may not support certain edge cases at this juncture. When the feature matures this message will be turned off.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================================================================================================================================================== short test summary info ==========================================================================================================================================================================================================
SKIPPED [1] tests/acceleration/test_acceleration_framework.py:164:  NOTE: this scenario will actually never happen, since in the code we always
    provide at least one dataclass (can consider to remove this test).
================================================================================================================================================================================================ 13 passed, 1 skipped, 27 warnings in 47.64s ================================================================================================================================================================================================
```

### CODE REVIEW

File: plugins/accelerated-peft/README.md [hunk 1]
```diff
@@ -19,9 +19,18 @@ Plugin | Description | Depends | Loading | Augmentation | Callbacks
 ## Known Issues
 
 - Models with sliding windows (e.g., Mistral, Mixtral) will have [memory and throughout issues](https://github.com/huggingface/transformers/issues/30461).
-- GPTQ-LORA sometimes observed to have `nan` grad norms in the begining of training, but training proceeds well otherwise.
+- GPTQ-LORA sometimes observed to have `nan` grad norms in the begining of training, but training pr.
```
Comment (fabianlim) [inline plugins/accelerated-peft/README.md:22]: something here is deleted by mistake.

File: plugins/accelerated-peft/README.md [hunk 2]
```diff
@@ -19,9 +19,18 @@ Plugin | Description | Depends | Loading | Augmentation | Callbacks
...
 - `low_cpu_mem_usage` temporarily disabled for AutoGPTQ until bug with `make_sure_no_tensor_in_meta_device` is resolved.
-- Requires nightly [AutoGPTQ](https://github.com/AutoGPTQ/AutoGPTQ) until package `> 0.7.1` becomes available
+- The legacy implementation of GPTQ-LORA uses an external AutoGPTQ package, you must ensure the package is installed
```
Comment (fabianlim) [inline plugins/accelerated-peft/README.md:24]: Lets not have this under Known Issues. Make a proper section in the README and explain
- the GPTQ depends on AutoGTPQ
- there are two options, using the package we extracted from Model Cloud
- or use the external lib
- external lib have to make sure to use this commit.

File: plugins/accelerated-peft/tests/test_peft_plugins.py [hunk 3]
```diff
@@ -26,6 +26,9 @@
     update_configuration_contents,
 )
 import pytest
+from unittest.mock import patch
+
+MODEL_NAME = "TheBloke/TinyLlama-1.1B-Chat-v0.3-GPTQ"
```
Comment (fabianlim) [inline plugins/accelerated-peft/tests/test_peft_plugins.py:31]: call this `MODEL_NAME_AUTO_GPTQ`

File: plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py [hunk 4]
```diff
@@ -48,12 +47,17 @@ def __init__(self, configurations: Dict[str, Dict], use_external_lib: bool = Fal
         self._check_config_equal(
             key="peft.quantization.auto_gptq.from_quantized", value=True
         )
-        self.use_external_lib = use_external_lib
+        self.use_external_lib = self._check_config_and_maybe_check_values(
```
Comment (fabianlim) [inline plugins/accelerated-peft/src/fms_acceleration_peft/framework_plugin_autogptq.py:50]: doing this will force a check that the config must have `external_lib`

Im not sure if `self._check_config_and_maybe_check_values(..., values=[True, False, None]` implements optional behavior. If this doesnt work, then maybe you can implement a `default` behavior

```
self._check_config_and_maybe_check_values(..., values=[True, False], default=False)
```

means that if the key is missing, it will set to False.

File: plugins/accelerated-peft/tests/test_peft_plugins.py [hunk 5]
```diff
@@ -86,6 +89,45 @@ def test_configure_gptq_plugin():
...
+                update_configuration_contents(
+                    read_configuration(CONFIG_PATH_AUTO_GPTQ),
+                    "peft.quantization.auto_gptq.use_external_lib",
+                    True,
+                ),
+                require_packages_check=False,
+            ) as framework:
+                pass
+
+    # - Test that gptqmodel is used when use_external_lib is False
```
Comment (fabianlim) [inline plugins/accelerated-peft/tests/test_peft_plugins.py:118]: have an extension of this test where
- we load `CONFIG_PATH_AUTO_GPTQ`  and delete away the key
- it will detect it as False
- and it will not use our gptq

File: plugins/accelerated-peft/tests/test_peft_plugins.py [hunk 6]
```diff
@@ -86,6 +89,45 @@ def test_configure_gptq_plugin():
...
+        update_configuration_contents(
+            read_configuration(CONFIG_PATH_AUTO_GPTQ),
+            "peft.quantization.auto_gptq.use_external_lib",
+            False,
+        ),
+        require_packages_check=False,
+    ) as framework:
+        model = framework.model_loader(MODEL_NAME)
+        assert any(isinstance(mod, QuantLinear) for mod in model.modules()), '''
+        use_external_lib=False, but local autogptq package not used by model'''
```
Comment (fabianlim) [inline plugins/accelerated-peft/tests/test_peft_plugins.py:130]:
```suggestion
        assert any(isinstance(mod, QuantLinear) for mod in model.modules()), \
           "use_external_lib=False, but local autogptq package not used by model"
```

**Review by fabianlim — CHANGES_REQUESTED** [review]
Looks in the right direction, but see comments

File: plugins/accelerated-peft/tests/test_peft_plugins.py [hunk 7]
```diff
@@ -86,6 +89,45 @@ def test_configure_gptq_plugin():
 
         e.match(f"AutoGPTQAccelerationPlugin: Value at '{key}'")
 
+def test_autogptq_loading():
+    def autogptq_unavailable(package_name: str):
+        return False
+
+    # - Test that error is thrown when use_external_lib is True but no package found.
```
Comment (fabianlim) [inline plugins/accelerated-peft/tests/test_peft_plugins.py:96]: it is best to leave very clear comments, I like this
1. Doing this
2. So that this is done
3. and this is checked..

style of comments https://github.com/foundation-model-stack/fms-hf-tuning/blob/9e4b49f25a7d9a383cd8e5dc1e3175e09197f890/tests/acceleration/test_acceleration_framework.py#L357

**Review by fabianlim — APPROVED** [review]
